compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -440,6 +440,8 @@ object Checking {
     }
     if (sym.isValueClass && sym.is(Trait) && !sym.isRefinementClass)
       fail(CannotExtendAnyVal(sym))
+    if (sym.isConstructor && !sym.isPrimaryConstructor && sym.owner.is(Trait, butNot = JavaDefined))
+      fail("Traits cannot have secondary constructors " + sym.owner.flagsString)
```

Comment:
Is `butNot = JavaDefined` necessary? Maybe add a doc for that.

Reply:
It is in the commit message

Reply:
It is in case a method is named `$init$` in a java interface.  It should not be considered as a constructor.